### DISCUSSION TITLE
Enable new key formats on gramine-ratls

### DISCUSSION THREAD

**Discussion by dzobbe** [gramineproject/gramine#1605] [Ideas]
Hi folks,

Just a minor thing. I have been struggling with Vault in letting it loading certs generated by `gramine-ratls`. At the end I discovered that it didn't like the enhanced cert format (i.e., "BEGIN TRUSTED CERTIFICATE"). I tried with DER but it didn't help. At the end, I discovered that by just changing the header (using the typical "BEGIN CERTIFICATE") Vault was able to load the certificate and the ratls-mbedtls client was also able to verify the quote. 
Maybe you can enable an option to generate certs by keeping the standard header.  

Cheers

**Comment by dimakuv** [comment]
According to this [SO answer](https://stackoverflow.com/a/52085885), Gramine's RA-TLS **must** generate the `BEGIN TRUSTED CERTIFICATE` line, because RA-TLS always generates **self-signed certificates**.

Gramine's RA-TLS uses mbedTLS to generate the self-signed X.509 certificate. The "self-signedness" is achieved by specifying the issuer that is the same as the subject: https://github.com/gramineproject/gramine/blob/d8d43440c040848699cb6e4bad606abe3ce39caa/tools/sgx/ra-tls/ra_tls_attest.c#L82-L84

IIUC, this logic forces mbedTLS to mark the certificate as "trusted certificate" and thus start the cert with `BEGIN TRUSTED CERTIFICATE`.

In other words, this is actually how it should be. This is not a bug, this is a feature.

Regarding your particular problem: a quick google search didn't give an answer on how to "load a self-signed certificate into Hashicorp Vault". What is the exact issue that Vault reports? Maybe you just need to add a parameter to the API that you use when loading the certificate; there must be something like `self-signed` or `insecure` argument I guess, which disables this Hashicorp's verification.

**Reply by dzobbe** [reply]
Sorry I forgot to reply @dimakuv . I am not saying it is a bug. Just saying that it would have been useful for situations where the BEGIN TRUSTED CERTIFICATE label creates issues. Vault returns this error:
```
Error initializing listener of type tcp: error loading TLS cert: tls: failed to find "CERTIFICATE" 
PEM block in certificate input after skipping PEM blocks of the following types: [TRUSTED CERTIFICATE]
```
I didn't find any solution for this except for changing the label.

**Reply by dimakuv** [reply]
@dzobbe Could it be a bug in Vault? Maybe you should post an issue on their GitHub (and link to our discussions here)?

I still don't think there is anything to be done on the Gramine/RA-TLS side.
